src/Snail/Lexer.hs

```diff
@@ -1,50 +1,99 @@
 module Snail.Lexer (
-  lexeme,
-  parens,
-  signedInteger,
-  skipBlockComment,
-  skipLineComment,
-  spaces,
-  symbol,
+  -- * Character lists
+  initialCharacter,
+  specialInitialCharacter,
+  digitCharacter,
+  specialSubsequentCharacter,
+
+  -- * The only parser you'll ever need
+  SExpression (..),
+  sExpression,
 ) where
 
+import Data.Functor (($>))
 import Data.Text (Text)
+import Data.Text qualified as Text
 import Data.Void
-import Text.Megaparsec
+import Text.Megaparsec hiding (Tokens (..))
 import Text.Megaparsec.Char
 import Text.Megaparsec.Char.Lexer qualified as L
 
 -- | TODO: 'Void' is the error type but we should use an explicit error type
 type Parser = Parsec Void Text
 
--- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin with that prefix
+{- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin
+ with that prefix
+-}
 skipLineComment :: Parser ()
 skipLineComment = L.skipLineComment ";"
 
--- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything in between
+{- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything
+ in between
+-}
 skipBlockComment :: Parser ()
 skipBlockComment = L.skipBlockCommentNested "#|" "|#"
 
--- | Generate a parser for whitespace in a language with 'skipLineComment' and 'skipBlockComment'
+{- | Generate a parser for whitespace in a language with 'skipLineComment' and
+ 'skipBlockComment'
+-}
 spaces :: Parser ()
 spaces = L.space space1 skipLineComment skipBlockComment
 
--- | A 'lexeme' is ...
-lexeme :: Parser a -> Parser a
-lexeme = L.lexeme spaces
-
 -- | Parse a 'Text' verbatim
 symbol :: Text -> Parser Text
 symbol = L.symbol spaces
 
--- | Parse an 'Integer' as a lexeme
-integer :: Parser Integer
-integer = lexeme L.decimal
-
--- | Parse a signed 'Integer'
-signedInteger :: Parser Integer
-signedInteger = L.signed spaces integer
-
 -- | Parse an S-Expression
 parens :: Parser a -> Parser a
 parens = between (symbol "(") (symbol ")")
+
+-- | A 'lexeme' is ...
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme spaces
+
+-- | ...
+initialCharacter :: String
+initialCharacter = ['a' .. 'z'] <> ['A' .. 'Z']
+
+-- | ...
+specialInitialCharacter :: String
+specialInitialCharacter = "!$%&*/:<=>?^_~"
+
+-- | ...
+digitCharacter :: String
+digitCharacter = ['0' .. '9']
+
+-- | ...
+specialSubsequentCharacter :: String
+specialSubsequentCharacter = "+-.@"
+
+-- | The list of valid token characters, note that we allow invalid tokens at this point
+validCharacter :: Parser Char
+validCharacter =
+  oneOf
+    ( initialCharacter
+        <> specialInitialCharacter
+        <> digitCharacter
+        <> specialSubsequentCharacter
+    )
+
+-- | ...
+data SExpression
+  = Token (SourcePos, Text)
+  | SExpression [SExpression]
+  deriving (Eq, Show)
```

Comment:
I don't believe this type of rose tree is implemented in a Haskell package

Reply:
Perhaps this: https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Tree.html#t:Tree

With `a ~ (SourcePos, Text)`, but you're probably better of defining a seperate data like you're doing here, and maybe provide a mapping/iso if you ever need it (I doubt it).

Reply:
Yeah, the tree in containers is nonempty so it doesn't quite work but I agree with the latter part.